isolate/mtn.go

```diff
@@ -0,0 +1,294 @@
+	c.Cfg.Url = cfg.Mtn.Url
+	c.Cfg.Headers = cfg.Mtn.Headers
+	return true
+}
+
+func (c *MtnState) PoolInit() bool {
+	if !c.Cfg.Enable {
+		return true
+	}
+	c.Pool = make(map[string]*IdState)
+	allAllocs, err := c.GetAllocations()
+	if err != nil {
+		// TODO: maybe need use main logger at this step
+		fmt.Println("%s ERROR: Cant init pool inside PoolInit(), err: %s", time.Now().UTC().Format(time.RFC3339), err)
+		return false
+	}
+	for netId, allocs := range allAllocs {
+		cState := *new(IdState)
+		cState.Reserved = 0
+		cState.Allocations = make(map[string]Allocation)
+		for _, alloc := range allocs {
+			cState.Allocations[alloc.Id] = Allocation{alloc.Net, alloc.Hostname, alloc.Ip, alloc.Id, false}
+		}
+		c.Pool[netId] = &cState
+	}
+	return true
+}
+
+func (c *MtnState) GetAllocations() (map[string][]Allocation, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 20 * time.Second)
+	defer cancel()
+	req, nrErr := http.NewRequest("GET", c.Cfg.Url + "?scheduler=" + c.Cfg.SchedLabel, nil)
+	if nrErr != nil {
+		return nil, nrErr
+	}
+	for header, value := range c.Cfg.Headers {
+		req.Header.Set(header, value)
+	}
+	req = req.WithContext(ctx)
+	rh, doErr := http.DefaultClient.Do(req)
+	defer rh.Body.Close()
+	if doErr != nil {
+		if rh.StatusCode == 400 {
+			errResp := AllocError{}
+			decoder := json.NewDecoder(rh.Body)
+			rErr := decoder.Decode(&errResp)
+			if rErr != nil {
+				return nil, fmt.Errorf(
+					"%s ERROR: Cant allocate. Request error: %s. Internal error: %s. Caused: %s.",
+					time.Now().UTC().Format(time.RFC3339),
+					doErr,
+					errResp.Message[0],
+					errResp.Cause.Message[0],
+				)
+			} else {
+				return nil, fmt.Errorf(
+					"%s ERROR: Cant allocate. Request error: %s. Body parse error: %s.",
+					time.Now().UTC().Format(time.RFC3339),
+					doErr,
+					rErr,
+				)
+			}
+		}
+		return nil, doErr
+	}
+	r := make(map[string][]Allocation)
+	jresp := []RawAlloc{}
+	decoder := json.NewDecoder(rh.Body)
+	rErr := decoder.Decode(&jresp)
+	if rErr != nil {
+		return nil, rErr
+	}
+	for _, a := range jresp {
+		r[a.Network] = append(r[a.Network], Allocation{a.Porto.Net, a.Porto.Hostname, a.Porto.Ip, a.Id, false})
+	}
+	return r, nil
+}
+
+func (c *MtnState) RequestAllocs(netid string) (map[string]Allocation, error) {
+	r := make(map[string]Allocation)
+	ctx, cancel := context.WithTimeout(context.Background(), 20 * time.Second)
+	defer cancel()
+	jsonBody := PostAllocreq{netid, c.Cfg.Ident, c.Cfg.SchedLabel}
+	txtBody, mrshErr := json.Marshal(jsonBody)
+	if mrshErr != nil {
+		return nil, mrshErr
+	}
+	for i := 0; i < c.Cfg.Allocbuffer; i++ {
+		req, nrErr := http.NewRequest("POST", c.Cfg.Url, bytes.NewBuffer(txtBody))
+		if nrErr != nil {
+			return nil, nrErr
+		}
+		for header, value := range c.Cfg.Headers {
+			req.Header.Set(header, value)
+		}
+		req.Header.Set("Content-Type", "application/json")
+		req = req.WithContext(ctx)
+		rh, doErr := http.DefaultClient.Do(req)
+		defer rh.Body.Close()
```

Comment:
err before defer and please, make sure u need defer inside a loop

Reply:
change that part, tnx